domain/cloudimagemetadata/state/state.go

```diff
@@ -0,0 +1,344 @@
+		var noTime time.Time
+		if m.CreationTime != noTime {
+			createdAt = m.CreationTime
+		}
+
+		mUUID, err := uuid.NewUUID()
+		if err != nil {
+			return errors.Annotatef(err, "failed to generate metadata uuid")
+		}
+		values = append(values, inputMetadata{
+			UUID:            mUUID.String(),
+			CreatedAt:       createdAt,
+			Source:          m.Source,
+			Stream:          m.Stream,
+			Region:          m.Region,
+			Version:         m.Version,
+			VirtType:        m.VirtType,
+			ArchitectureID:  archId,
+			RootStorageType: m.RootStorageType,
+			RootStorageSize: m.RootStorageSize,
+			Priority:        m.Priority,
+			ImageID:         m.ImageId,
+		})
+	}
+
+	// Prepare statement to insert a batch of image in db. In case of conflict on unique constraint,
+	// it will update the image id instead
+	insertMetadataStmt, err := sqlair.Prepare(`
+INSERT INTO cloud_image_metadata (*)
+VALUES ($inputMetadata.*)
+ON CONFLICT (stream, region, version, architecture_id, virt_type, root_storage_type, source) DO UPDATE
+SET image_id = EXCLUDED.image_id;`, inputMetadata{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return errors.Trace(db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// Insert images metadata
+		return errors.Trace(tx.Query(ctx, insertMetadataStmt, values).Run())
+	}))
+}
+
+// DeleteMetadataWithImageID deletes all metadata associated with the given image ID from the database.
+func (s *State) DeleteMetadataWithImageID(ctx context.Context, imageID string) error {
+	db, err := s.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	deleteMetadataStmt, err := sqlair.Prepare(`
+DELETE FROM cloud_image_metadata 
+WHERE image_id = $metadataImageID.image_id`, metadataImageID{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return errors.Trace(db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, deleteMetadataStmt, metadataImageID{imageID}).Run()
+		return err
+	}))
+}
+
+// FindMetadata retrieves cloud image metadata from the database based on specified filter criteria.
+// It constructs a dynamic SQL query using the supplied criteria and executes it to fetch matching records.
+// Returns [cloudmetadataerrors.NotFound] if none are found with this criteria.
+func (s *State) FindMetadata(ctx context.Context, criteria cloudimagemetadata.MetadataFilter) ([]cloudimagemetadata.Metadata, error) {
+	db, err := s.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// clauses will collect all required clauses to build the final WHERE ... AND ... clause
+	clauses := []string{
+		// Ignores expired metadata
+		fmt.Sprintf(`cloud_image_metadata.created_at >=  '%s'`, time.Now().Add(-ExpirationDelay).Format(time.DateTime)),
+	}
+
+	// declareEqualsClause is an helper function to add a clause of the type field = 'value'
+	declareEqualsClause := func(field string, value string) {
+		if value != "" {
+			clauses = append(clauses, fmt.Sprintf(`%s = '%s'`, field, value))
```

Comment:
These are incorrect. Don't build the strings up based on user input. Instead, build the strings as `? = ?` and then pass the ordered arguments to the query.

Reply:
Oh my gosh, you right. Shame on me.

I am not sure that sqlair support `? = ?` but I will definetely fix that possible source of injection. Thanks!